solver/src/solver/cow_dex_ag_solver.rs

```diff
@@ -0,0 +1,644 @@
+use crate::solver::solver_utils::Slippage;
+use crate::solver::zeroex_solver::api::SwapQuery;
+use crate::solver::zeroex_solver::api::SwapResponse;
+use crate::solver::zeroex_solver::api::ZeroExResponseError;
+use crate::solver::zeroex_solver::STANDARD_ZEROEX_SLIPPAGE_BPS;
+use crate::solver::{Auction, Solver};
+use crate::solver::{ParaswapSolver, ZeroExSolver};
+use crate::{liquidity::LimitOrder, settlement::Settlement};
+use anyhow::{anyhow, Result};
+use contracts::GPv2Settlement;
+use ethcontract::{Account, H160, U256};
+use futures::future::join_all;
+use reqwest::Client;
+use shared::{token_info::TokenInfoFetching, Web3};
+use std::collections::HashMap;
+use std::sync::Arc;
+
+pub struct CowDexAgSolver {
+    pub account: Account,
+    pub paraswap: ParaswapSolver,
+    pub zeroex: ZeroExSolver,
+}
+
+pub struct CowDexAgSolverParameters {
+    pub account: Account,
+    pub web3: Web3,
+    pub settlement_contract: GPv2Settlement,
+    pub token_info: Arc<dyn TokenInfoFetching>,
+    pub slippage_bps: u32,
+    pub disabled_paraswap_dexs: Vec<String>,
+    pub client: Client,
+    pub partner: Option<String>,
+    pub chain_id: u64,
+}
+
+impl CowDexAgSolver {
+    pub fn new(parameters: CowDexAgSolverParameters) -> Self {
+        CowDexAgSolver {
+            account: parameters.account.clone(),
+            paraswap: ParaswapSolver::new(
+                parameters.account.clone(),
+                parameters.web3.clone(),
+                parameters.settlement_contract.clone(),
+                parameters.token_info,
+                parameters.slippage_bps,
+                parameters.disabled_paraswap_dexs,
+                parameters.client.clone(),
+                parameters.partner,
+            ),
+            zeroex: ZeroExSolver::new(
+                parameters.account,
+                parameters.web3,
+                parameters.settlement_contract,
+                parameters.chain_id,
+                parameters.client,
+            )
+            .unwrap(),
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl Solver for CowDexAgSolver {
+    // todo: Solver does not yet deal with unwrap and warps
+    async fn solve(&self, Auction { orders, .. }: Auction) -> Result<Vec<Settlement>> {
+        if orders.is_empty() {
+            return Ok(Vec::new());
+        }
+        // Randomize which orders we start with to prevent us getting stuck on bad orders.
+        // orders.shuffle(&mut rand::thread_rng())
+        let mut orders: Vec<LimitOrder> = orders.into_iter().collect();
+        // for now, lets only solve for up to 5 orders
+        orders.truncate(5);
```

Comment:
Why limit the amount of orders? Shouldn't we be limiting the amount of pairs instead?